rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,842 @@
+focusables and specify navigation method within itself.
+
+A [`MenuSetting`] can be used to:
+* Prevent navigation from one specific submenu to another
+* Specify if 2d navigation wraps around the screen.
+* Specify "scope menus" such that a
+  `NavRequest::ScopeMove` emitted when
+  the focused element is a [`Focusable`] nested within this `MenuSetting`
+  will navigate this menu.
+* Specify _submenus_ and specify from where those submenus are reachable.
+* Specify which entity will be the parents of this [`MenuSetting`], see
+  `MenuSetting::reachable_from` or `MenuSetting::reachable_from_named` if you don't
+  have access to the `Entity`
+  for the parent [`Focusable`]
+
+If you want to specify which [`Focusable`] should be
+focused first when entering a menu, you should mark one of the children of
+this menu with `Focusable::prioritized`.
+
+#### Invariants
+
+**You need to follow those rules  to avoid panics**:
+1. A menu must have **at least one** [`Focusable`] child in the UI
+   hierarchy.
+2. There must not be a menu loop. I.e.: a way to go from menu A to menu B and
+   then from menu B to menu A while never going back.
+
+#### Panics
+
+Thankfully, programming errors are caught early and you'll probably get a
+panic fairly quickly if you don't follow the invariants.
+* Invariant (1) panics as soon as you add the menu without focusable
+  children.
+* Invariant (2) panics if the focus goes into a menu loop.
+
+#### Fields
+
+```rust
+pub struct MenuSetting {
+    /// Whether to wrap navigation.
+    ///
+    /// When the player moves to a direction where there aren't any focusables,
+    /// if this is true, the focus will "wrap" to the other direction of the screen.
+    pub wrapping: bool,
+    /// Whether this is a scope menu.
+    ///
+    /// A scope menu is controlled with [`NavRequest::ScopeMove`]
+    /// even when the focused element is not in this menu, but in a submenu
+    /// reachable from this one.
+    pub scope: bool,
+}
+```
+
+#### Going back to a previous menus
+
+`Focusable`s have a `prioritized` state that is set when they go from
+active/focused to not active.
+This is a form of memory that allows focus to go back to the last focused
+element within a menu when it is re-visited.
+This is also the mechanism used to let the user decide which focusable to focus
+when none are focused yet.
+
+#### Spawning menus
+
+Menus store a reference to their parent,
+but that parent is not necessarily their hierarchical parent.
+The parent is just a button in another menu.
+
+It is inconvenient to have to pre-spawn each button to acquire their `Entity` id
+just to be able to spawn the menu you'll get to from that button.
+[`bevy-ui-navigation`] uses a proxy holding both the parent
+and the menu state info.
+
+That proxy component is `MenuSeed`,
+you can initialize it either with the `Name` or `Entity` of the parent focusable of the menu.
+A system will take all `MenuSeed` components and replace them with a `TreeMenu`.
```

Comment:
Note that the current implementation has simplified this. `MenuSetting` is now a component, and we have another `MenuBuilder` component that still gets converted by the navigation plugin into a `TreeMenu`.

Issue is I don't like "`MenuBuilder`" and wait to have a better name before updating the RFC.